### PR TITLE
Implement Claw interface

### PR DESCRIPTION
**Pull request by arporter**
This PR is for work done under #114.

### CODE REVIEW

**Comment by arporter** [comment]
While starting work on introducing OpenACC directives with CLAW, I discovered test failures in this branch. I don't understand why. Maybe I've updated the version of various things on my laptop. The most problematic failure was to do with losing the stderr from where I 'shell out' to run CLAW. I've fixed that but accidentally committed it to the wrong branch so I've used `git cherry-pick` to apply it to the branch for this PR.

**Comment by arporter** [comment]
I made the mistake of working on (and so breaking) my branch of CLAW that this PR depends upon. I've now cut a new `psyclone` branch in stfc/claw-compiler which should be used with this PR.

File: doc/claw.rst [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+The version of the CLAW compiler used by PSyclone is available from
+https://github.com/stfc/claw-compiler.  See the INSTALL.md file for
+installation instructions.  Once it is installed, you will need to
+manually copy the Python interface module into the `lib` directory of
+the Claw installation:
+
+::
+
+    cp claw-compiler/omni-cx2x/src/claw/python/ClawTransform.py <claw-install-root>/lib
```
Comment (rupertford) [inline doc/claw.rst]: I could not find a file called ClawTransform.py or even a directory called omni-cx2x.

Comment (arporter) [reply doc/claw.rst]: My apologies - this goes back to the fact that Valentin has re-structured CLAW and I got carried-away and pulled his changes into the stfc version of CLAW. The file you need is now:

    cx2t/src/claw/python/ClawTransform.py

(On the `psyclone` branch of stfc/claw-compiler)

**Comment by rupertford** [comment]
I have a potentially disruptive suggestion. In your implementation the PSyclone script passes a list of kernels and a script file. What I propose is for CLAW transformations to be called in the same way PSyclone transformations are called. So, for example

```
def trans(psy):
    import claw
    ...
    kernel = psy......
    ...
    claw_openacc_trans = claw.trans.openacckern()
    new_kernel = claw_openacc_trans.apply(kernel)
    ...
```

Do you think this is a) feasible and, if so b) worth doing now in this pr?

**Comment by arporter** [comment]
I think it is feasible and probably even sensible for specific cases such as the one your example suggests - that of adding the necessary OpenACC `routine` directive to a kernel. However, I'm concerned that in general, this would be equivalent to requiring PSyclone to support a set of transformations that operate on a whole PSy layer, rather than letting the user build-up their recipe from fundamental transformations (such as loop fuse). For specific, simple cases it could be powerful but I think, in general, allowing users to write their own script is better. e.g. if we think of transforming the matrix-vector kernel in the various ways that Karthee/Sergi have done, then that would need a script (I think) rather than a one-size fits all transformation.

**Comment by rupertford** [comment]
Ah, I see your point. The PSyclone script, as it stands, does not know about the innards of a Kernel.

Given that you are writing a script that does know about the innards of a Kernel (a "CLAW" script, rather than a "PSyclone" script) and given that both are just Python, is there any particular reason for formally separating the two i.e. is there any reason why the contents of a "CLAW" script could not be in a "PSyclone" script?

**Comment by arporter** [comment]
Nope, there's no reason and in fact I have written the examples so that both the PSyclone and CLAW scripts are routines in the same file (we have a `trans()` and a `claw_trans()`). I thought it best to separate the two routines as one is run from CLAW and only operates on a CLAW AST whereas the original works on our PSyclone schedule.

File: doc/getting_going.rst [hunk 2]
```diff
@@ -60,9 +60,13 @@ example in openSUSE 42.2:
 
 and then use pip2.7 instead of pip.
 
+PSyclone supports the use of the CLAW compiler
+(https://github.com/C2SM-RCM/claw-compiler) to perform kernel
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/getting_going.rst]: Done.

**Comment by rupertford** [comment]
Andy and I have discussed the limitations of the current approach to integrating CLAW into PSyclone i.e. the way it has been implemented in this PR. The fact that we need to use jython, which only works for Python2, that we have to separate out scripts working on CLAW and scripts working on the PSyclone IR and that it is difficult to make use of existing CLAW functionality by only driving CLAW via a script are quite strong limitations.

We feel that a potentially better approach would be for PSyclone to drive CLAW by adding directives to Kernel code, then running CLAW. The Python transformation interface could remain, it would just be interfacing with CLAW in a different way. The advantages of this approach is that we don't need to modify CLAW with a Python scripting interface, we don't need to use Jython, we can mix PSyclone and CLAW transformations in whatever way we like and we can potentially make use of existing CLAW functionality e.g. optimizing for a GPU as a black-box high-level optimization. We can also have access to the OMNI XML Fortran representation allowing us to modify code outside of CLAW if required, e.g. when inlining.

If @arporter is happy I'll therefore close this pull request and add the above information to the original issue.

**Comment by rupertford** [comment]
Closing this pull request as we expect to use a different approach

**Pull request closed without merging**